torch/csrc/autograd/engine.cpp

```diff
@@ -723,27 +771,48 @@ auto Engine::execute(const edge_list& roots,
 }
 
 void Engine::enqueue_blocked_task_on_cpu(NodeTask task) {
-  std::call_once(start_threads_flag_, &Engine::start_threads, this);
-  ready_queue(at::kCPU).push(
+  std::call_once(start_device_threads_flag_, &Engine::start_device_threads, this);
+  std::shared_ptr<GraphTask> graph_task = task.base_.lock();
+  TORCH_INTERNAL_ASSERT(graph_task, "GraphTask is no longer valid!");
+  ready_queue(graph_task, at::kCPU).push(
       std::move(task), /* incrementOutstandingTasks */ false);
 }
 
 std::shared_ptr<FutureVariableList> Engine::execute_with_graph_task(
     const std::shared_ptr<GraphTask>& graph_task,
     std::shared_ptr<Node> graph_root) {
-  std::call_once(start_threads_flag_, &Engine::start_threads, this);
+  std::call_once(start_device_threads_flag_, &Engine::start_device_threads, this);
   // Lock mutex for GraphTask.
   std::unique_lock<std::mutex> lock(graph_task->mutex_);
 
-  ready_queue(at::kCPU).push(
+  ready_queue(graph_task, at::kCPU).push(
       NodeTask(graph_task, std::move(graph_root), InputBuffer(0)));
 
-  // Not a worker
+  // worker_device == NO_DEVICE it's a CPU thread and it's trying to drive the
+  // autograd engine with corresponding GraphTask, and its NOT a re-entrant call
   if (worker_device == NO_DEVICE) {
     // graph_task_exec_post_processing is done when the Future is marked as
     // completed in mark_graph_task_completed.
+
+    // We set the worker_device to CPU_DEVICE only if worker_device was previously
+    // NO_DEVICE. Setting it to CPU afterwards allow us to detect whether this is
+    // a re-entrant call or not.
+    //
+    // If worker_device = NO_DEVICE: this is not a re-entrant call
+    // If worker_device is other devices (i.e. CPU, CUDA): this is a re-entrant
+    //    backward call from that device.
+    worker_device = CPU_DEVICE;
+
+    // set the graph_task owner to the current device
+    graph_task->owner_ = worker_device;
+
+    // The owning thread start to drive the engine execution with the GraphTask
+    // that has already been pushed to the current CPU thread's ready_queue
+    lock.unlock();
+    thread_main(nullptr, false);
```

Comment:
Is this going to block until the entire GraphTask execution is done? If so, this is a problem since `execute_with_graph_task` is supposed to be nonblocking and should return a Future as soon as possible.

Currently, `execute_with_graph_task` is called from the RPC threadpool in distributed autograd, if this call blocks until all execution for the GraphTask is done, it will likely cause a deadlock for distributed autograd.

Reply:
Yeah this is going to block the computation until GraphTask finishes, it's because the owning thread is right now itself a worker thread, and it need to do the backward computation by itself rather than idling and waiting for results. I am thinking of how this could integrate with distributed autograd..

Reply:
Well, clearly, if you want the work to happen in a nonblocking fashion, CPU has to run somewhere else. I wonder if we should just spawn another CPU thread on the fly in the non-blocking case.

Reply:
we can use the main c10 thread pool for non-blocking case (the one shared with fork/join). We basically can just enqueue task with the body of this function and return the future

Reply:
@dzhulgakov I'm assuming you mean executing `thread_main` here in the JIT threadpool. Note that `thread_main` is going to block on IO for distributed autograd. As a result, I don't think it would be a good idea to run it within the JIT threadpool. Since as far as I understand, JIT threads don't wait on any IO today and this would break the assumption.

Also, note that we plan to merge the RPC and the JIT threadpool and if we do that, we could easily cause a deadlock as follows:

1. We launch 4 backward passes on a JIT threadpool of size 4.
2. All of these backward passes execute a bit and then wait on IO. Note that now all JIT threads have been used up and are just idling.
3. We are now deadlocked since RPC requests coming in to unblock these backward passes can't be executed since all JIT threads have been used up.

Reply:
@pritamdamania87  There has been discussion about long-term merging thread pools between different use cases (e.g. rpc / JIT / autograd cpu), etc. but one explicit pre-condition is going to be first eliminating blocking calls within those code paths.

Reply:
yeah use the c10 threadpool could potentially cause deadlock as @pritamdamania described, we might can assume in short term distributed autograd only handle CPU training without cross device condition, and make distributed autograd use the non-blocking/continuation logic. Then we could aim at long term solution.